### PR TITLE
Slice Number Sweep: 128, 144 vs baseline 96

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The Transolver architecture divides the input point cloud into `slice_num=96` attention "slices" for routing features. The Gap/Stagger Spatial Bias (GSB) mechanism — our key confirmed win — works by modulating slice routing via learned spatial biases conditioned on gap and stagger.

**Hypothesis:** More slices = more fine-grained routing paths = better geometric specialization for OOD tandem configurations. The GSB can only route samples into 96 distinct attention patterns at slice_num=96 — increasing this allows the model to develop more specialized representations for different gap/stagger combinations, improving transfer to NACA6416.

The additional VRAM cost is modest (slice attention scales with slice_num, not slice_num²), and with 96GB GPUs we have headroom. Training time may increase ~5-10%.

## Instructions

Run **2 seeds (42, 73)** for **each of 2 configs**, using `--wandb_group tanjiro-slice-sweep`:

**Config A — 128 slices:**
```bash
cd cfd_tandemfoil && python train.py --agent tanjiro \
  --wandb_name "tanjiro/slice128" --wandb_group tanjiro-slice-sweep \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 128 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --pcgrad_3way --pcgrad_extreme_pct 0.15 --gap_stagger_spatial_bias \
  --seed 42
```
(repeat with `--seed 73`)

**Config B — 144 slices:**
```bash
cd cfd_tandemfoil && python train.py --agent tanjiro \
  --wandb_name "tanjiro/slice144" --wandb_group tanjiro-slice-sweep \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 144 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --pcgrad_3way --pcgrad_extreme_pct 0.15 --gap_stagger_spatial_bias \
  --seed 42
```
(repeat with `--seed 73`)

Run all 4 seeds in parallel. Report p_in, p_oodc, p_tan, p_re for each config (2-seed averages).

**Important:** If VRAM issues arise at slice_num=144, reduce batch size first, or fall back to just config A (128).

## Baseline

| Metric | Baseline (2-seed avg) |
|--------|----------------------|
| p_in | 13.05 |
| p_oodc | 7.70 |
| **p_tan** | **28.60** |
| p_re | 6.55 |

Baseline W&B runs: d7l91p0x (seed 42, p_tan=28.9), j9btfx09 (seed 73, p_tan=28.3)

**Reproduce baseline:**
```bash
cd cfd_tandemfoil && python train.py --agent <name> --wandb_name "<name>/baseline-gsb-pcgrad" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --pcgrad_3way --pcgrad_extreme_pct 0.15 --gap_stagger_spatial_bias
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT tanjiro:

## Results

All 4 experiments completed (181 min each, timeout-limited).

### Per-run results

| Run | Epoch | val_loss | p_in | p_tan | p_oodc | p_re | W&B |
|-----|-------|----------|------|-------|--------|------|-----|
| slice128-s42 | 145 | 0.3866 | 13.6 | 29.2 | 8.0 | 6.5 | 245dka5p |
| slice128-s73 | 145 | 0.3894 | 13.5 | 29.9 | 7.7 | 6.6 | aasc7vel |
| slice144-s42 | 140 | 0.4037 | 15.3 | 29.6 | 8.2 | 6.9 | pz5vo3h0 |
| slice144-s73 | 139 | 0.3982 | 14.2 | 29.8 | 8.3 | 7.0 | 1wzqtn7w |

### 2-seed averages vs baseline (slice_num=96)

| Config | p_in | p_tan | p_oodc | p_re | Δ p_tan |
|--------|------|-------|--------|------|---------|
| **Baseline (slice=96)** | **13.05** | **28.60** | **7.70** | **6.55** | — |
| slice=128 | 13.55 | 29.55 | 7.85 | 6.55 | +3.3% worse |
| slice=144 | 14.75 | 29.70 | 8.25 | 6.95 | +3.8% worse |

### Analysis

**Hypothesis rejected.** Both slice=128 and slice=144 are worse than baseline slice=96 across all metrics.

**slice=128** (p_tan=29.55): Consistent with the previous PR #2155 result (slice=128 → p_tan 29.50 avg). The additional GSB signal doesn't help with more slices — 96 remains optimal.

**slice=144** (p_tan=29.70): Even worse, and clearly degraded on p_in (14.75 vs 13.05, +13%). The larger model hit fewer epochs (139-140 vs 145 for slice=128) due to the timeout, meaning it had less time to converge. The per-epoch cost is higher, and the returns are negative.

**Key observations:**
1. **Convergence speed matters.** slice=144 only reached epoch 139-140 vs 145 for slice=128 and ~156 for slice=96. The additional parameters require more epochs to converge, but the 180-min timeout limits total training.
2. **More slices = more parameters ≠ better generalization.** With GSB already providing explicit geometry conditioning, extra routing capacity just adds noise — the model has more degrees of freedom to overfit routing without improving tandem transfer.
3. **p_in degradation scales with slice count.** slice=128 p_in=13.55 (+3.8%), slice=144 p_in=14.75 (+13%). More slices hurt in-distribution performance too, suggesting overfitting to routing patterns.
4. **This confirms the PR #2155 finding:** 96 slices is the sweet spot for this architecture+data combination. Both previous (64, 96, 128) and current (128, 144) sweeps point to the same conclusion.

### VRAM

All runs completed without OOM on dedicated GPUs. No batch size reduction needed.

### Suggested follow-ups

- **Slice count is confirmed at 96.** No further sweeps needed.
- Instead of more slices, improving GSB itself may be more productive: e.g., a deeper MLP for the spatial bias computation (2 hidden layers instead of 1), or adding foil-2 DSDF features to the spatial bias input.
- The p_tan gap (28.60 vs target) likely needs loss-level or data-level innovations rather than routing capacity changes.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2171 because ADVISOR: Closing — both configs worse. slice=128: p_tan +3.3%, slice=144: p_tan +3.8%.

Confirms slice_num=96 is optimal (also validated by prior PR #2155). More slices means:
1. More parameters → more routing overfitting
2. Fewer epochs in 180-min window (144 only reached epoch 139-140)
3. GSB already provides explicit geometry conditioning; extra routing capacity is noise

The scaling pattern is clear: p_in degradation scales with slice count (128: +3.8%, 144: +13%). The model has more freedom to overfit routing without improving transfer.

Dead end: slice_num changes. 96 confirmed optimal by multiple experiments.

Good systematic work..
